### PR TITLE
Remove redundant `found_inf` recompute from `_step_supports_amp_unscaling` path

### PR DESCRIPTION
**Pull request by crcrpar**
following https://github.com/pytorch/pytorch/pull/97415#issuecomment-1499787115.

Rel: https://github.com/pytorch/pytorch/pull/98613

cc @vincentqb @jbschlosser @albanD @janeyx99 @mcarilli @ptrblck @leslie-fang-intel @jgong5 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
This PR does make all paths consistent in that they would check for nonsense _before_ unscaling, but I'm still not super convinced this is the way to go. This looks to be BC breaking (@ngimel @albanD to confirm) so would deserve some documentation.

**Review by janeyx99 — COMMENTED** [review]
I do agree this is a more viable path after the hf t5 discussion. I would expect this to have better loss/performance now in the benchmarks--could you do a benchmark run and post the results?

**Review by janeyx99 — APPROVED** [review]

**Comment by crcrpar** [comment]
*** Results with nan/inf filter:

| Variation                          |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:-----------------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused --fp16 0 |                              317.64 |           8 |            2.21 |
| --optim adamw_torch_fused --fp16 1 |                              366.33 |          25 |            2.55 |
| --optim adamw_torch --fp16 0       |                              293.59 |           0 |            2.21 |
| --optim adamw_torch --fp16 1       |                              363.73 |          24 |            2.55 |

```

*** Setup:


Datetime    : 2023-04-17 16:18:17

Software:
transformers: 4.29.0.dev0
torch       : 2.1.0a0+git6e1e27f
cuda        : 12.1
python      : 3.10.9

Hardware:
1 GPUs      : NVIDIA A100 80GB PCIe, 79.18GB


*** The benchmark command line was:

python scripts/benchmark/trainer-benchmark.py --base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 32 ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch' '--fp16 0|--fp16 1' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

*** Results without nan/inf filter:

| Variation                          |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:-----------------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused --fp16 0 |                              329.45 |           8 |            2.21 |
| --optim adamw_torch_fused --fp16 1 |                              381.84 |          25 |          nan    |
| --optim adamw_torch --fp16 0       |                              304.38 |           0 |            2.21 |
| --optim adamw_torch --fp16 1       |                              375.60 |          23 |          nan    |

```

*** Setup:


Datetime    : 2023-04-17 16:13:00

Software:
transformers: 4.29.0.dev0
torch       : 2.1.0a0+git6e1e27f
cuda        : 12.1
python      : 3.10.9

Hardware:
1 GPUs      : NVIDIA A100 80GB PCIe, 79.18GB


*** The benchmark command line was:

python scripts/benchmark/trainer-benchmark.py --base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 32 \
--no_logging_nan_inf_filter ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch' '--fp16 0|--fp16 1' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

**Comment by janeyx99** [comment]
Is it correct to understand "nan/inf filtering" to mean the nan/inf check for unscaled gradients (aka the check after unscaling)? Or does it refer to a different thing entirely? 

It is odd to me that the nans show up for the fused but not the unfused...

**Comment by crcrpar** [comment]
"nan/inf filtering' refers to https://huggingface.co/docs/transformers/v4.28.1/en/main_classes/trainer#transformers.TrainingArguments.logging_nan_inf_filter: "Whether to filter nan and inf losses for logging. If set to True the loss of every step that is nan or inf is filtered and the average loss of the current logging window is taken instead."



> It is odd to me that the nans show up for the fused but not the unfused...

I'm afraid I think you misread the table. There are two rows whose Train loss is nan. One is "--optim adamw_torch_fused --fp16 1" and the other is "--optim adamw_torch --fp16 1". Both use `GradScaler`, the optimizer of the former is `AdamW(..., fused=True)` and that of the latter is `AdamW(...)`.

**Comment by janeyx99** [comment]
Yea, this makes sense, it's as I was suspecting in https://github.com/pytorch/pytorch/issues/96755#issuecomment-1511929203. 

I think we can land this PR then!

File: torch/cuda/amp/grad_scaler.py [hunk 1]
```diff
@@ -336,6 +336,8 @@ def step(self, optimizer, *args, **kwargs):
             # and `found_inf` to the passed optimizer so that the optimizer can utilize those
             # to skip the parameter updates or unscale gradients before updating parameters in
             # the fused kernel, e.g. `FusedAdamMathFunctor`.
+            if optimizer_state["stage"] is OptState.READY:
+                self._check_inf_per_device(optimizer)
             kwargs_ = kwargs
             has_grad_scaler_kwarg = "grad_scaler" in inspect.signature(optimizer.step).parameters
             if has_grad_scaler_kwarg:
```
Comment (janeyx99) [inline torch/cuda/amp/grad_scaler.py:343]: Just confirming--is it right to say that this branch does not need updates because other optimizers (like the apex ones) already handle the check_inf/nan order for the scaled grads only?

Comment (crcrpar) [reply torch/cuda/amp/grad_scaler.py:343]: updates of what?

if `optimizer._step_supports_amp_scaling` is `True`, then the optimizer takes care of both unscaling gradients and parameter updates

Comment (albanD) [reply torch/cuda/amp/grad_scaler.py:343]: But it shouldn't re-check again if the unscaling already happened right?

Comment (crcrpar) [reply torch/cuda/amp/grad_scaler.py:343]: this if statement is to avoid the unwanted re-check of gradients. `GradScaler.unscale_(optimizer)` updates `OptState` to `OptState.UNSCALED`: https://github.com/pytorch/pytorch/blob/d8d479c854c622a6ef21190b7e62c9a76f0ea2a7/torch/cuda/amp/grad_scaler.py#L285

Comment (janeyx99) [reply torch/cuda/amp/grad_scaler.py:343]: As long as the future optimizers that would fall under the has_grad_scaler_kwarg already reuse the calculated found_inf dictionary instead of recomputing, this would be true. 

I was asking for a confirmation because I am not familiar with the implementations of those optimizers, though it looks like we don't define a contract on our side.

Comment (crcrpar) [reply torch/cuda/amp/grad_scaler.py:343]: ah, would you mean `_check_inf_per_device` should not be called if `has_grad_scaler_kwarg` because we have no control in that path?

Comment (janeyx99) [reply torch/cuda/amp/grad_scaler.py:343]: It depends on what our contract is, though it looks like it hasn’t been defined previously. I did a GH search and saw that apex fused adam would check the infs in spite of the unscale_ api (see https://github.com/NVIDIA/apex/blob/817e8187fc3960985af89bd09094f08082307080/apex/optimizers/fused_adam.py#L204). 

So I guess it does not really matter right now where the check inf goes, though it looks more correct to be where you’ve put it now (under the specific branch). I’m not sure if we want to establish the contract that the grad scaler will set the state to UNSCALED vs READY and have optimizers condition on those states. Since that path is already in deprecation stages, I do not think it is high priority. Feel free to land this!

**Comment by janeyx99** [comment]
this would be a PR to include in our release 2.0.1 once merged

**Comment by albanD** [comment]
> this would be a PR to include in our release 2.0.1 once merged

Cherry pick period is over AFAIK

**Comment by janeyx99** [comment]
Oh oops I misread the dev discuss timeline--nvm about 2.0.1 then!

**Comment by ngimel** [comment]
Cherry-pick I think is extended to 4/21 (double check with @atalman) to accommodate mha changes for 4090

**Comment by crcrpar** [comment]
Is ["sebotnet33ts_256                    FAIL:     accuracy=fail_accuracy, expected=pass"](https://github.com/pytorch/pytorch/actions/runs/4745660281/jobs/8428788904?pr=98620#step:13:998) real?

with a brief search I saw some pull requests that either mark this case as flaky or loosen the threshold

**Comment by janeyx99** [comment]
@pytorchbot merge -ic

@crcrpar it’s existed on trunk, I see it still appearing on trunk even now https://hud.pytorch.org/minihud?name_filter=inductor, kicking off a merge job ignoring that failure.
